apps/examples/uart_hw_test/uart_hw_test.c

```diff
@@ -0,0 +1,246 @@
+ * Private functions
+ ****************************************************************************/
+static int poll_read(int fd, void *buf, int len)
+{
+#ifndef CONFIG_DISABLE_POLL
+	struct pollfd fds[1];
+	fds[0].fd = fd;
+	fds[0].events = POLLIN;
+	int ret = poll(fds, 1, UART_POLL_TIMEOUT_MS);
+	if (ret > 0) {
+		return read(fd, buf, len);
+	} else if (ret == 0) {
+		printf("uart poll read: TIMEOUT reading data on uart\n");
+		return 0;
+	}else {
+		printf("uart poll read: ERROR reading data on uart : %d\n", ret);
+		return -1;
+	}
+#else
+	return read(fd, buf, len);
+#endif
+}
+
+static void uart_hw_rx_test(int fd)
+{
+	int i = 0;
+	uint8_t ch;
+	char buf[5] = {'\0'};
+
+	printf("test_uart_rx: Test START\n\r");
+
+	/* Loop until we receive the start value */
+	do {
+		if (poll_read(fd, &ch, 1) <= 0) {
+			return;
+		}
+	} while (ch != UART_TEST_START_VAL);
+
+	/* Loop until we receive the stop value */
+	do {
+		if (poll_read(fd, &ch, 1) <= 0) {
+			return;
+		}
+		if (ch == UART_TEST_VAL) {
+			i++;
+		}
+	} while (ch != UART_TEST_STOP_VAL);
+
+	/* Return the count of bytes received without error */
+	itoa(i, buf, 10);
+	write(fd, buf, strlen(buf));
+
+	printf("test_uart_rx: Test COMPLETE\n\n\r");
+}
+
+static void uart_hw_tx_test(int fd)
+{
+	char ch;
+
+	printf("test_uart_tx: Test START\n\r");
+
+	/* Send test start value */
+	ch = UART_TEST_START_VAL;
+	write(fd, &ch, 1);
+
+	/* Send test value for "cnt" times */
+	ch = UART_TEST_VAL;
+	for (int i = 0; i < UART_TEST_COUNT; i++) {
+		write(fd, &ch, 1);
+	}
+
+	/* Send stop value to indicate end of test */
+	ch = UART_TEST_STOP_VAL;
+	write(fd, &ch, 1);
+
+	printf("test_uart_tx: Test COMPLETE\n\n\r");
+}
+
+static void configure_tty(int fd, int baud)
+{
+	struct termios ttyio;
+	memset(&ttyio, 0, sizeof(ttyio));
+
+	tcgetattr(fd, &ttyio);
+
+	/* set baudrate */
+	cfsetispeed(&ttyio, baud);
+	cfsetospeed(&ttyio, baud);
+
+	ttyio.c_cflag &= ~CSIZE;
+	ttyio.c_cflag |= CS8;
+	ttyio.c_cflag &= ~CSTOPB;
+	ttyio.c_cflag &= ~(PARENB | PARODD);
+
+	tcflush(fd, TCIOFLUSH);
+	tcsetattr(fd, TCSANOW, &ttyio);
+}
+
+static void run_background_console_logs(void)
+{
+	while (1) {
+		char ch = 'a';
+		while (ch <= 'z') {
+			putchar(ch++);
+			usleep(10);
+		}
+	}
+}
+
+/****************************************************************************
+ * Public functions
+ ****************************************************************************/
+#ifdef CONFIG_BUILD_KERNEL
+int main(int argc, FAR char *argv[])
+#else
+int uart_hw_test_main(int argc, char **argv)
+#endif
+{
+	int fd = 0;
+	int p = 0;
+	char port[20] = {'\0'};
+	char ch = 0 ;
+	int bauds[] = {115200, 38400, 19200, 9600, 8400};
+
+	printf("Enter UART port number\n");
+	p = getchar();
+	p = atoi((const char *)&p);
+
+	sprintf(port, UART_DEV_PATH, p);
+	fd = open(port, O_RDWR, 0666);
+	if (fd < 0) {
+		printf("ERROR: Failed to open UART port : %s\n", port);
+		return -1;
+	}
+
+	for (int i = 0; i < sizeof(bauds) / sizeof(int); i++) {
```

Comment:
how does tash app and ubuntu app sync start and end for each baudrate?

Reply:
Added handshake mechanism.